corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +12,60 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
         return value
     else:
         return None
+
+
+class UserChangeLogger(object):
+    def __init__(self, domain, user, is_new_user, changed_by_user, changed_via):
+        self.domain = domain
+        self.user = user
+        self.is_new_user = is_new_user
+        self.changed_by_user = changed_by_user
+        self.changed_via = changed_via
+
+        self.original_user_doc = self.user.to_json()
+        self.fields_changed = {}
+        self.messages = []
+
+    def add_changes(self, changes):
+        # ignore for new user since the whole user doc is logged for a new user
+        if self.is_new_user:
+            return
+        for name, new_value in changes.items():
+            if self.original_user_doc[name] != new_value:
+                self.fields_changed[name] = new_value
+
+    def add_change_message(self, message):
+        # ignore for new user since the whole user doc is logged for a new user
+        if self.is_new_user:
+            return
+        self.messages.append(message)
+
+    def add_info(self, info):
+        # useful info for display like names
+        self.messages.append(info)
+
+    def save(self):
+        self._include_user_data_changes()
+        if self.is_new_user:
+            log_user_change(
+                self.domain,
+                self.user,
+                changed_by_user=self.changed_by_user,
+                changed_via=self.changed_via,
+                message=". ".join(self.messages),
+                action=UserModelAction.CREATE,
+            )
+        else:
+            log_user_change(
+                self.domain,
+                self.user,
+                changed_by_user=self.changed_by_user,
+                changed_via=self.changed_via,
+                message=". ".join(self.messages),
+                fields_changed=self.fields_changed
+            )
+
+    def _include_user_data_changes(self):
```

Comment:
It looks that would already be covered, since
1. profile is updated as a part of user_data
2. and we track any changes to user_data [here](https://github.com/dimagi/commcare-hq/blob/f3950d7abea9aa7e833b531e35e2d30936d195ae/corehq/apps/user_importer/helpers.py#L63)
Though i see now that since that is happening in logger and not in importer because the original doc is available with the logger, it's hidden, which is not great. Going to see and move it to the importer instead

Reply:
https://github.com/dimagi/commcare-hq/pull/29935/commits/55958b87cbd5df577474352d4bbae32497794f7f

Reply:
The existing `user_data` handling will only show the profile id, won't it? It'd be helpful to be able to see the name.

Reply:
Thank you. 

https://github.com/dimagi/commcare-hq/pull/29935/commits/983593f86efc0c926bfe44e716365ec448462242

It was a bit confusing that the implementation would look for profile name under "commcare_profile" column but one could still update user data by passing an ID for profile and that is what tests are using. 
So i tried to cover both cases for logging.

Reply:
Sounds like the original tests aren't realistic, which is probably my bad.